### ISSUE TITLE
Discrepancy Between Balancing Loss in Paper and Code Implementation

### ISSUE THREAD

**Issue by DAVID-NGUYEN-S16** [RobertCsordas/moeut#4]
I am confused about the balancing loss formula in the paper, which differs from the implementation in MoEUT.  

![Image](https://github.com/user-attachments/assets/2da434a8-1566-4d04-8ecd-28402e574b2f)  

In Equation 4, the loss function takes inputs $\( x_t \)$ and $\( W_S \)$, but in the code, the input is transformed as follows:  

```python
def entropy_reg(sel: torch.Tensor, dim: int) -> torch.Tensor:
    sel = F.log_softmax(sel, dim=-1)
    sel = log_mean(sel, dim)
    return - entropy_l(sel).mean()
```

Can you clarify why this difference exists?

**Comment by RobertCsordas** [comment]
x_t W_s is the selection code used also for selecting the experts. The code feeds the already computed scores instead of recomputing them. The argument "sel" in is x_t W_s. The score itself is calculated here: https://github.com/RobertCsordas/moeut/blob/87f04973d3db3ad7cb4ca67a0750241e40210d3c/moeut/moeut.py#L88, and saved for regularization later.

Let me know if this clarifies your question.

**Comment by DAVID-NGUYEN-S16** [comment]
I understand it, but I'm confused about why you use log softmax and log mean before computing entropy. I don't see you reporting anything about it. When I read the paper, I understand that the input for entropy is just softmax(x @ W).

**Comment by RobertCsordas** [comment]
Oh, that is the same, except it is in log space. Handling probabilities in log space tends to be more numerically stable. entropy_l(x) calculates - sum_i x[i] e^x[i], so what comes out of that function is the standard entropy.
